Basic tutorial (#1)
flame-engine/flame_tutorials: tutorials/basic_1_square/README.md

# Basic: Rendering a simple square on the screen

This tutorial will introduce you to:

 - `Game`: The basic class to build a game on, you will understand its structure and how to use it to build a simple game
 - `GameWidget`: The widget used to place your game on your widget tree
 - Basic rendering: You will get introduced to the basics of using `dart:ui` class `Canvas` which is the class which Flame uses for rendering your game.

> All the code of this tutorial code can be found [here](./code) and is based on Flame `1.0.0-rc7` version.

Reviewer: rc8?

Author: Fixed!

By the end of this tutorial you will have built a simple game that renders a square bouncing on the screen that will look like this:

![Preview](./media/preview.gif)

## Building your first game

> `Game` is the most basic class that you can use to build your game, it includes the necessary methods for creating a basic game loop and methods for the lifecycle of a Flame game.

Reviewer 2: Maybe it would be better to use BaseGame for the most basic tutorial since BaseGame is what most people will want to use in their game later?

Reviewer 3: Agreed, even tho the `Game` class is indeed the most basic class, it is not the easiest of them to use when you are just learning Flame. As most of our own documentation is more focused on both FCS and `BaseGame`.

`Game` is a more a low-level API class imho

Reviewer 3: But a counter to that would be that setting up FCS without basic understanding of our gameloop/logic is also counter-intuitive.. So I can go either way on this

Author: I think we should approach Game before approaching BaseGame. I intentionally choose Game here, because I think I think it is important that people know that there are both classes and why they exist and most important, I think that for this tutorial, which introduces what a game loop is, this is way more fit than BaseGame.

I intent to do more tutorials afterwards and basically all of them will use BaseGame instead of Game, just these firsts one I will use Game.

Reviewer: I would at least comment: normally you will use BaseGame that contains a lot of utilities but we will see that on tutorial xxx

Author: I don't think we should mention `BaseGame` yet, IMO it will only take the attention of the reader elsewhere

Reviewer 2: I agree that a comment about BaseGame would be good too, I don't think the user will drift away because of that.

Author: Alright, added a comment about BaseGame, let me know if that is fine

> `GameWidget` is, like its name suggest, a Flutter widget that will run your game and place it inside the Flutter widget tree.

Reviewer: like the name suggests

> As a first step, lets just put a Game instance running, for that we will need to create our own class that extends Flame's Game class, implement its methods, and pass an instance of that Game to a GameWidget, somehting like this:

Reviewer 3: ```suggestion
As a first step, lets just put a Game instance running, for that we will need to create our own class that extends Flame's Game class, implement its methods, and pass an instance of that Game to a GameWidget, something like this:
```

Also I dont think the start of this sentence is correct? "Lets put just put a Game instance running,". Not sure what you meant with that.

> As a first step, lets just put a Game instance running, for that we will need to create our own class that extends Flame's Game class, implement its methods, and pass an instance of that Game to a GameWidget, somehting like this:

Reviewer 3: I would also like to suggest that when we reference things like classes(`Game`, `GameWidget`) that we use backticks to make sure that they are words related to code. To keep it more consistent overal.

> As a first step, lets just put a Game instance running, for that we will need to create our own class that extends Flame's Game class, implement its methods, and pass an instance of that Game to a GameWidget, something like this:

Reviewer: get a Game instance running. For that, we will

Reviewer: to a GameWidget. Something like this:

Author: Fixed!

```dart
import 'package:flutter/material.dart';
import 'package:flame/game.dart';

void main() {
  runApp(GameWidget(

>     game: MyGame(),

Reviewer 3: Is it not standard to put the creation of a `Game` into it's own variable, to ensure it wont be accidentally rebuilt? So that people will be more prone to do that later on as well?

Author: I will put it on its own variable but since this is directly in the `main` function, that is not a problem.

>   ));

Reviewer: missing trailing comma

Author: Done 🙄

}

class MyGame extends Game {
  @override
  void update(double dt) { /* TODO */ }

>  }

Reviewer: this is wrong

Author: what is wrong? why is it wrong?

Author: Can't spot anything wrong here.

Reviewer: it is definitely wrong, this line should not exist

Author: It is definitely not wrong, just to make sure I just copied and pasted this line on dartpad and there is no errors here:

![Captura de Tela 2021-03-17 às 10 05 31](https://user-images.githubusercontent.com/835641/111473058-61c9c280-8709-11eb-86f0-1b3ead75c519.png)

  @override
  void render(Canvas canvas) { /* TODO */ }
}
```

> That is it, you should see just an empty black screen for now, but now we have the bare bones needed to start implementing our game.

Reviewer: That is it! If you run this, you should see...

Reviewer: you will see might be even better

Author: changed!

Before going further, it is important to explain what those two methods mean.

> Flame's `Game` class is an implementation of a Game Loop, which the basic structure on which basically all games are built on. It is called Game Loop because it really works a an infinite loop that continues to iterate as long as the game is running. This loop can be divided on some steps:

Reviewer 3: Technically speaking the `Game` class itself is not the Game loop right? I would just say that the `Game` class provides a game loop with which the basic structure can be defined.

> Flame's `Game` class is an implementation of a Game Loop, which the basic structure on which basically all games are built on. It is called Game Loop because it really works a an infinite loop that continues to iterate as long as the game is running. This loop can be divided on some steps:

Reviewer 3: ```suggestion
Flame's `Game` class is an implementation of a Game Loop, which the basic structure on which basically all games are built on. It is called a Game Loop because it really works as an infinite loop that will continue to iterate as long as the game is running. This loop can be categorized as followed:
```

> Flame's `Game` class is an implementation of a Game Loop, which the basic structure on which basically all games are built on. It is called a Game Loop because it really works as an infinite loop that will continue to iterate as long as the game is running. This loop can be categorized as followed:

Reviewer: basically all games -> most games (so we don't get any angry comments)

Author: Don't see how we could get angry comments, basically all games uses a game loop.

But I changed either way

Reviewer: refer to audioplayers, there is always a way!

 - 1: Take input from the player
 - 2: Process the logic for the next frame
 - 3: Render the frame.

> On this tutorial we will focus the steps two and three.

Reviewer 3: ```suggestion
In this tutorial we will focus on both step two and three.
```

> To process the logic of the game, we use the `update` method, which is ran always before the frame is rendered and receives a single argument, a `double` value called `dt` (delta time), which is the ammout os seconds from the current iteration and the last one. This delta time is very important so we can always know how to correct calculate movement, animations and etc.

Reviewer 2: Amount of between
Correctly

Maybe add something about that it is the speed of movement, animations etc that it will keep correct?

Reviewer: missing oxford comma

Author: I have no idea what an oxford comma is

Author: @spydon fixed the first typos, abou the explanation on speed, I did that on the end of the paragraph, is that not enough? Or would you state it differently?

Reviewer 2: It just says "correctly calculate..." in the last paragrapgh, not how dt makes the difference, that is what I think would be good to have there, an explanation that it is about the speed.

Author: Oh, I see, I have added a better explanation further on the tutorial, on the comment explaining what is happening on the update method. I think it covers what you meant here. Please take a look on it and let me know if that is enough, otherwise we can think on a better way to explain that.

> To render the frame, we use the `render` method, this method receives a single argument which is an instance of a `dart:ui` `Canvas` class. With that instance we can basically render anything we want. It is important to not that in this method, we should only contain render logic, any game logic should be put in the `update` method.

Reviewer 2: It is missing a word in the last sentence after "to not"

Author: Fixed, I changed the sentence a little bit

> 

Reviewer: the way I like to explain this is to define the world state, which is a static snapshot of all the information needed by your game. then you can define:

update: (state, dt) -> next state
render: (state) -> image on screen

we an also add a diagram like the ones @renancaraujo makes

Author: Not sure we need that here, I have added an explanation on what is happening on each instruction on the update method later on the tutorial

Reviewer: I think it's important for people that never coded any games before, the game loop might be a unfamiliar concept

Author: Yeah, but my point is that it is already being explained later, I am not sure how to expand more on that. Maybe you send me the explanation you want to be here?
Or even better you could do a follow up here later? What do you think?

Reviewer: I will read it again and comment if I miss anything

> Now that we have a better understanding of how the game structures works, lets start to plan our game.

Reviewer 3: ```suggestion
Now that we have a better understanding of how the game structure works, lets start to plan our game.
```

> First things first, lets think on what variables and data structure we would need. For that, lets recap what we are building: A simple game where a square keeps bouncing forever from one side of the screen to the other. Thinking on this, we will need:

Reviewer: this is already the middle/end of the tutorial, weird to say First things first here.

Author: Changed to a simple `So,`, let me know if you have any other suggestions

 - A variable to tell us the speed of the square in logical pixels per second.
 - A variable to inform which direction the square is moving.

>  - A structure to represent our square, which has a position and dimensions.

Reviewer: might be good to make distinction here between static data (speed, dimension) and actual world state (direction, position), to not confuse readers

Author: Done

> With that in mind, check the example below, note the comments for explanations on which code section.

Reviewer 2: On each code section?

Author: I always confuse which and each, thanks!

```dart
class MyGame extends Game {
  // A constant speed, represented in logical pixels per second

>   static const int square_speed = 400;

Reviewer: snake_case?!?

Author: Done

  // To represent our square we are using the Rect class from dart:ui

>   // which is a handle class to represent this type of data. We will be

Reviewer: handle class? what is that?

Author: I was going to handy, fixed

  // seeing other types of data classes in the future, but for this
  // example, Rect will do fine for us.

>   late Rect squarePos;

Reviewer: you could also use just a Vector2, which might be simpler to do some operations (you can have another one, static, for the dimensions)

Author: For this tutorial, I find Rect better, will probably introduce Vector2 when we start explaining other features

  // To reprent our direction, we will be using an int value, where 1 means
  // going to the right, and -1 going to the left, this may seems like a too much
  // simple way of representing a direction, and indeed it is, but this will
  // will work fine for our small example and will make more sense when we implement
  // the update method

>   int squareDirection = 1;

Reviewer: you could also use as data the 1d velocity if that makes it more clear than a sign variable

Author: not sure what you mean by 1d velocity

Reviewer: signed speed

Author: I still have no idea what is the suggestion here

Reviewer: to bundle the information of direction and speed into a single variable representing the 1d velocity as a signed double. but it's just an idea, I like the current implementation also and use both a lot. just trying to think what is simpler to understand for the reader. probably doesn't even make a difference.

  // The onLoad method is where all of the game initialization is supposed to go
  // for this example, you may think that this square could just be initialized on the field

>   // declaration, and you are right, but for ludic reasons and to present the life cycle method

Reviewer 2: Ludic reasons?

Author: I just looked on the dictionary and this word don't have the meaning that I though it did. I was going for something like "to teach you about this method".

Any suggestions here? Maybe "but for learning purposes"?

Author: changed to learning purposes

  // for this example we will be initializing this field here.
  @override
  Future<void> onLoad() async {
    squarePos = Rect.fromLTWH(0, 0, 100, 100);
  }

  // Update and render ommited
}
```

Right, now we have all the data and variables we need to start implementing our game, for the next step, lets just draw our little square on the screen.

```dart
  @override
  void render(Canvas canvas) {

>     // Canvas is a class from dart:ui and is reponsible for all the renderizations on Flame

Reviewer 2: All the rendering

Author: fixed

    //

>     // BasicPalette is a help class from Flame, which provides default, pre builded instances

Reviewer 2: Pre-built

Author: fixed

    // of Paints that can be used by your game

>     canvas.drawRect(squarePos, BasicPalette.white.paint);

Reviewer: static variable

Author: I have no idea what you are suggesting here

Reviewer: extract the paint to a static variable to avoid creating things every tick

Author: Made a comment on the previous place, I think we should solve that on Flame for not creating paint every time this is accessed since this is easily cacheable

  }
}
```

You may now be seeing a static white square being rendered on the top left corner of your screen, which is not that cool, so, to finish our example, lets add some movement to the game and implement our update method:

```dart
  @override
  void update(double dt) {

>     // Here we move our square by applying our iteration delta time (dt) to our

Reviewer: I think we can double down here on the fact that the update method must update your game state to the next state after an amount of time dt seconds. if you dont consider dt the game will run differently depending on hardware/software conditions

Author: Added a new section for this

    // speed variable and direction. Note that the Rect class class is immutable
    // and the translate method returns a new Rect instance for us, so we just
    // re assign it to our square variable
    squarePos = squarePos.translate(square_speed * squareDirection * dt, 0);

>     // This simple verifies if the square is going right, and has reached the end of the screen

Reviewer 2: Simple condition?

Author: Done

    // if so, we just invert our direction

>     if (squareDirection == 1 && squarePos.right > size.x) {

Reviewer: probably good to callout that size is the logical size of the canvas (if not using viewports/camera)

Author: Done

      squareDirection = -1;
    // This does the same, but now checking the left direction
    } else if (squareDirection == -1 && squarePos.left < 0) {
      squareDirection = 1;
    }
  }
```

> And that is it for this basic tutorial, here we have covered the basics of Flame, its basic classes and some basic rendering. From that we can start to build more complex things and more "exciting" games.

Reviewer: basic tutorial, in which we have

basic basic basic

I dont think we need to quote exciting, sounds a bit patronizing

Author: Fixed first and last.

Any suggestions for basic basic basic?

Reviewer: up to you. its not a really big issue anyway
